core/src/main/java/com/graphhopper/routing/weighting/custom/ExpressionBuilder.java

```diff
@@ -0,0 +1,382 @@
+import org.codehaus.janino.Scanner;
+import org.codehaus.janino.*;
+import org.codehaus.janino.util.DeepCopier;
+import org.locationtech.jts.geom.prep.PreparedGeometryFactory;
+
+import java.io.*;
+import java.util.*;
+import java.util.concurrent.atomic.AtomicLong;
+
+class ExpressionBuilder {
+    private static final AtomicLong longVal = new AtomicLong(1);
+    private static final String AREA_PREFIX = "in_area_";
+    private static final Set<String> allowedNames = new HashSet<>(Arrays.asList("edge", "Math"));
+    private static final boolean JANINO_DEBUG = Boolean.getBoolean(Scanner.SYSTEM_PROPERTY_SOURCE_DEBUGGING_ENABLE);
+
+    // TODO without a cache we get X% slower routing and Y% slower routingLM8. CH requests and preparation is unaffected
+    // as cached weighting from preparation is used. This cache ensures that the first Weighting classes, typically the
+    // ones specified in the profiles, are never removed regardless of how frequent other Weightings are created and accessed.
+    // We only need to synchronize the get and put methods alone. E.g. we do not care for the race condition where
+    // two identical classes are requested and one of them is overwritten. We could use CachingJavaSourceClassLoader from
+    // Janino but 1. we would need to use a single compiler across threads and 2. the statements (for priority and speed)
+    // are still unnecessarily created.
+    private static final int CACHE_SIZE = Integer.getInteger("graphhopper.custom_weighting.cache_size", 10);
+    // TODO perf compare with ConcurrentHashMap, but I guess, if there is a difference at all, it is not big for small maps
+    private static final Map<String, Class> CACHE = Collections.synchronizedMap(new HashMap<>(CACHE_SIZE));
+
+    // Introduce a dynamic cache to remember different Weighting classes, but throw away less frequently used classes.
+    // Use accessOrder==true to remove oldest accessed entry, not oldest inserted.
+    private static final int DYN_CACHE_SIZE = Integer.getInteger("graphhopper.custom_weighting.dynamic_cache_size", 1000);
+    private static final Map<String, Class> DYN_CACHE = Collections.synchronizedMap(
+            new LinkedHashMap<String, Class>(DYN_CACHE_SIZE, 0.75f, true) {
+                protected boolean removeEldestEntry(Map.Entry eldest) {
+                    return size() > DYN_CACHE_SIZE;
+                }
+            });
+
+    /**
+     * This method compiles a new subclass of CustomWeightingHelper composed from the provided CustomModel caches this
+     * and returns an instance.
+     */
+    static CustomWeightingHelper create(CustomModel customModel, EncodedValueLookup lookup,
+                                        double globalMaxSpeed, double maxSpeedFallback, DecimalEncodedValue avgSpeedEnc) {
+        Java.CompilationUnit cu;
+        try {
+            String key = customModel.toString() + ",global:" + globalMaxSpeed + ",fallback:" + maxSpeedFallback;
```

Comment:
StringBuilder?

Reply:
I do not like StringBuilder. It makes everything less readable and JIT is able to optimize string concatenation into it if "hot code". And in this case here I'm pretty sure it is not "hot code".